server/ec2spotmanager/tasks.py

```diff
@@ -131,566 +113,247 @@ def check_instance_pool(pool_id):
-            smallest_size = instance_size
-            smallest = [instance_type]
-        elif instance_size == smallest_size:
-            smallest.append(instance_type)
-    # replace the allowed instance types with those that are <= count, or the smallest if none are
-    config.ec2_instance_types = acceptable_types or smallest
-
-    # Figure out where to put our instances
-    try:
-        (region, zone, instance_type, rejected) = _get_best_region_zone(config)
-    except (boto.exception.EC2ResponseError, boto.exception.BotoServerError, ssl.SSLError, socket.error):
-        # In case of temporary failures here, we will retry again in the next cycle
-        logger.warning("[Pool %d] Failed to acquire spot instance prices: %s.", pool.id, traceback.format_exc())
-        return
-    except RuntimeError:
-        logger.error("[Pool %d] Failed to compile userdata.", pool.id)
-        entry = PoolStatusEntry()
-        entry.type = POOL_STATUS_ENTRY_TYPE['config-error']
-        entry.pool = pool
-        entry.isCritical = True
-        entry.msg = "Configuration error: %s" % traceback.format_exc()
-        entry.save()
-        return
-
-    priceLowEntries = PoolStatusEntry.objects.filter(pool=pool, type=POOL_STATUS_ENTRY_TYPE['price-too-low'])
-
-    if not region:
-        logger.warning("[Pool %d] No allowed region was cheap enough to spawn instances.", pool.id)
-
-        if not priceLowEntries:
-            entry = PoolStatusEntry()
-            entry.pool = pool
-            entry.type = POOL_STATUS_ENTRY_TYPE['price-too-low']
-            entry.msg = "No allowed region was cheap enough to spawn instances."
-            for zone in rejected:
-                entry.msg += "\n%s at %s" % (zone, rejected[zone])
-            entry.save()
-        return
-    else:
-        if priceLowEntries:
-            priceLowEntries.delete()
-
-    # convert count from cores to instances
-    #
-    # if we have chosen the smallest possible instance that will put us over the requested core count,
-    #   we will only be spawning 1 instance
-    #
-    # otherwise there may be a remainder if this is not an even division. let that be handled in the next tick
-    #   so that the next smallest instance will be considered
-    #
-    # eg. need 12 cores, and allow instances sizes of 4 and 8 cores,
-    #     8-core instance costs $0.24 ($0.03/core)
-    #     4-core instance costs $0.16 ($0.04/core)
-    #
-    #     -> we will only request 1x 8-core instance this time around, leaving the required count at 4
-    #     -> next time around, we will request 1x 4-core instance
-    count = max(1, count // CORES_PER_INSTANCE[instance_type])
-
-    logger.info("[Pool %d] Using instance type %s in region %s with availability zone %s.",
-                pool.id, instance_type, region, zone)
-
+    cloud_provider = Provider().getInstance(pool.config.provider)
+    if isinstance(pool.config.provider, EC2PoolConfiguration):
```

Comment:
The whole point of the abstraction layer is to eliminate this kind of per-provider `if`.

Reply:
this has been removed.